src/vdaf.rs

```diff
@@ -57,7 +58,9 @@ pub enum VdafError {
 }
 
 /// An additive share of a vector of field elements.
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Clone, Debug)]
+// Only derive equality checks in test code, as the content of this type is a secret.
+#[cfg_attr(feature = "test-util", derive(PartialEq, Eq))]
```

Comment:
I'm concerned that this API might be too permissive. The intent is to fence the unsafe comparison to testing, but IIUC, the user can still shoot themself in the foot if they're not careful. Let's say I enable "test-util" in my crate that I also deploy to production; then I get to use `Eq` in my production code, right?

Reply:
One way we could resolve this is to write `ConstantTimeEq` implementations for the relevant structs/enums, and write `PartialEq` implementations that dispatch to them. (for an example of this pattern, see `IdpfPublicShare`) This would provide safe APIs for both downstream dependents and for test code, avoid the need for the `test-util` Cargo feature, and still give us the convenience benefit of being able to use `assert_eq!()`.

Reply:
That sounds reasonable to me.

Reply:
The concern that led to this implementation was (from https://github.com/divviup/libprio-rs/issues/722):
> But the risk there is that new VDAFs will use derived PartialEq implementations that aren't constant time. So perhaps we would want to remove all PartialEq + Eq trait bounds and instead define some custom ConstantTimeEq trait.

That is, if we expose `PartialEq + Eq` by default (and expect users of the library to use these traits for comparison), we risk future VDAFs implementing variable-time versions of these traits. This would lead to any DAP implementations built on these VDAFs silently becoming insecure, in a way that would be particularly hard to notice.

OTOH, I like the idea of safe-by-default implementations and especially not requiring callers to know about `subtle` at all to use the library. I have moved everything relevant over to a constant-time implmementation of `PartialEq + Eq`.


> Let's say I enable "test-util" in my crate that I also deploy to production; then I get to use Eq in my production code, right?

Correct; if you enable a feature called "test-util" in your production binaries, IMO you are responsible for what comes next. :)